rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+
+---
+
+The two cases above highlight a weakness of using semver in our environment, as we care about identifying forward but not backwards compatible changes, and the semantic version does not traditionally capture this. As such we explored a potential solution by tracking two versions.
+
+### Combined Versions
+
+A potential solution to the problems caused by the above is to store a pair of semantic versions. If we specify type versions as the _ordered_ pair `(P, C)`, where `P` is the semantic version according to the producer, and `C` is the semantic version according to the consumer then both sides are able to track compatibility.
+
+Example:
+
+A new type:
+
+```json
+{
+  "P": "1.0.0",
+  "C": "1.0.0"
+}
+```
+
+An optional field is added to the Type
+
+```json
+{
+  "P": "1.1.0",
+  "C": "2.0.0"
+}
+```
+
+An optional field is removed from the Type
+
+```json
+{
+  "P": "2.0.0",
+  "C": "2.1.0"
+}
+```
+
+This allows us to compare the versions depending on the circumstance outlined above, where we're trying to track forwards or backwards compatibility. In addition to being a rather unituitive and confusing solution, this still has a few large problems.
+
+### Other Problems
+
+- As outlined above, we start with a type, add an optional field, and then remove an optional field. This _should_ theoretically be the same type as the previous one (outside of versioning info), although semver tells us that there was a major change between them, and as such we are not able to infer compatibility from the versioning information. _Because of this, we'll almost always need to check compatibility regardless_
+- Updating a **major** version of a deeply nested type, for instance a data type, will require **major** version changes to all types using it (if they wish to update). This is deeply in contrast to most uses of semver like versioning libraries. In those cases, major versions of sub-dependencies often cause the need to update some internal logic, but these breaking changes to not propagate up the dependency chain. If we used semver for types, we'd sometimes see propagating major version changes, causing very large major versions in the higher-level types with lots of dependencies. This is a problem because the major appeal of semantic versioning would be the assumptions an implementation could make about compatibility in non-major version differences.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [Semantic Versioning](https://semver.org)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How should we constrain valid URI formats
+  - How should the version pair be formatted? Should we use query parameters?
+  - Should it be possible to go to a URI _without_ expressing a version, if so should returning the latest version be the expected behavior
```

Comment:
Potentially this is something we don't need to specify as part of the RFC and leave up to the implementer, but I think we should do this yes (i.e. return the latest version of the type when no version is specified)

Reply:
I agree that I think we should. I also think we _probably_ do want to include it in the RFC so that a Type Nexus on blockprotocol.org could rely on the presence of that path.

Reply:
[W3C use](https://www.w3.org/2005/05/tr-versions) the root for "the most mature Recommendation of a given technology or the most agreeing-on version". They also use `latest` for the last stable version of a doc, effectively the same thing. In their model `upcoming` denotes a version of a document with WIP changes (which we could use for unreleased edits).

[HASH Core](https://hash.ai/platform/core) uses `stable` for the last cut release of a thing, and its root URL serves _either_ this (if a user does not have edit permissions), or the tip containing unreleased changes (if a user has edit permissions).

I think the W3C model is probably clearer, where root = stable = latest release (with or without a redirect), and another specified keyword can be used to view unreleased edits (possibly depending on permissions managed by an EA).

Reply:
I can add a section _recommending_ that hosts also serve `/v/latest` to redirect to (or directly serve) the most recent published version of a type. I don't think we want to **enforce** it at this point yet, because I'm not sure of the ramifications yet for hosting the URIs on file-systems, etc. If it was clearly very valuable and something we needed for programmatic functionality then it makes sense to require it, but at the moment it seems a nice to have.

I'd also be inclined to not talk about things like `upcoming`, I don't think "unreleased edits" or unstable versions, etc. are particularly useful in the context of schemas, which are much more constrained in behaviour than something like a library or npm package.

Reply:
Despite literally just writing otherwise.. perhaps we do want to provision for unstable/beta/whatever versions of types 🙃 It might be helpful to host temporary URIs for the purposes of development, so a Block Schema can be unstable alongside an unstable release of the Block for beta testing / etc.

Although if we go down that road I think we'd have to spend quite a while thinking about the implications and making sure we don't introduce problems. Probably worth us not provisioning for them now, and instead adding to the spec when we have a better understanding of how this stuff works in practice.

Reply:
I've added a section talking about `/latest`